### PR TITLE
Spotlight — highlight search keyword

### PR DESCRIPTION
**Pull request by igorlesnenko**
#5545

https://user-images.githubusercontent.com/466991/143566195-bcefd471-f8ab-4490-b90f-1844ea13f4b4.mov

**How to test**

- Open spotlight
- Type anything in a search field
- See matched keyword highlighted
- If the matched reflection card is deep inside a reflection group, that card should pop up on the top at the stack

### CODE REVIEW

**Comment by igorlesnenko** [comment]
@tianrunhe can you review please?

File: packages/client/components/TaskEditor/decorators.ts [hunk 1]
```diff
@@ -12,11 +13,34 @@ const findEntity = (entityType) => (contentBlock, callback, contentState) => {
...
+    return
+  }
+
+  const text = contentBlock.getText()
+  const textLower = text.toLowerCase()
+  const searchQueryLower = searchQuery.toLowerCase()
+
+  let start = 0
+  let found = true
+  while (found) {
```
Comment (nickoferrall) [inline packages/client/components/TaskEditor/decorators.ts:27]: This is probably just personal preference, but I find `find` methods to be more readable than `while` loops. Would something like this work?
```
textLower.split('').findIndex((letter) => searchQueryLower.includes(letter))
```

File: packages/client/hooks/useEditorState.ts [hunk 2]
```diff
@@ -3,10 +3,11 @@ import {convertFromRaw, EditorState} from 'draft-js'
 import {useEffect, useRef, useState} from 'react'
 import makeEditorState from '../utils/draftjs/makeEditorState'
 import mergeServerContent from '../utils/mergeServerContent'
+import remountDecorators from '../utils/draftjs/remountDecorators'
 
-const useEditorState = (content?: string | null | undefined) => {
+const useEditorState = (content?: string | null | undefined, searchQuery?: string | null) => {
```
Comment (nickoferrall) [inline packages/client/hooks/useEditorState.ts:8]: Ideally, I think `useEditorState` shouldn't need to know about the `spotlightSearchQuery`. I'm not too familiar with DraftJS, and I've always found it to be a pain when dealing with it, but I think we can set state with the decorator. See here: https://draftjs.org/docs/advanced-topics-decorators/#setting-new-decorators

File: packages/client/components/ReflectionGroup/ReflectionGroup.tsx [hunk 3]
```diff
@@ -135,6 +136,8 @@ const ReflectionGroup = (props: Props) => {
             dragUserId
             isSpotlight
           }
+          content
```
Comment (nickoferrall) [inline packages/client/components/ReflectionGroup/ReflectionGroup.tsx:139]: Is `content` being used in this file?

File: packages/client/components/TaskEditor/decorators.ts (hunk 1 above)
Comment (igorlesnenko) [reply packages/client/components/TaskEditor/decorators.ts:27]: @nickoferrall Here, I need to find all occurrences `startindex` and `endIndex` without overlapping and call a callback function for each found.

Not sure I got the idea with `split` and `findIndex`. 
With `split` I see kind of following solution:

```js
textLower.split(searchQueryLower).map(function (token) {
    return this.pos += token.length + searchQueryLower.length
}, { pos: -searchQueryLower.length }).slice(0, -1).forEach(i => { 
  callback(i, i + searchQueryLower.length)
})
```

What do you think?

File: packages/client/components/ReflectionGroup/useSpotlightVisibleReflections.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    if (spotlightSearchQuery != null && visibleReflections.length > 1) {
+      const spotlightSearchQueryLower = spotlightSearchQuery.toLowerCase()
+      const matchIndex = visibleReflections.findIndex((reflection) => {
+        const textLower = reflection.plaintextContent.toLowerCase()
+        return textLower.indexOf(spotlightSearchQueryLower) !== -1
+      })
+
+      if (matchIndex > 0) {
+        visibleReflections.unshift(visibleReflections.splice(matchIndex, 1)[0])
```
Comment (nickoferrall) [inline packages/client/components/ReflectionGroup/useSpotlightVisibleReflections.ts:20]: NIT: I'd find it easier to read if the splice was moved out of the unshift function. Something like: 
```
const matchingReflection = visibleReflections.splice(matchIndex, 1)[0]
visibleReflections.unshift(matchingReflection)
```

File: packages/client/components/TaskEditor/decorators.ts (hunk 1 above)
Comment (nickoferrall) [reply packages/client/components/TaskEditor/decorators.ts:27]: Ahh OK, thanks for clarifying! If we can avoid making `findText` a thunk, I would be very happy as I find them quite tricky. That might just be me so feel free to ignore this suggestion and stick with your original version, but could this work?

```
  if (searchQuery) {
    const findMatchingText = (
      contentBlock: ContentBlock,
      callback: (startIdx: number, endIdx: number) => void
    ) => {
      const textLower = contentBlock.getText().toLowerCase()
      const searchQueryLower = searchQuery.toLowerCase()
      const matchingIdx = textLower.indexOf(searchQueryLower)
      if (matchingIdx !== -1) {
        callback(matchingIdx, matchingIdx + searchQueryLower.length)
      }
    }

    compositeDecorator.push({
      strategy: findMatchingText,
      component: SearchHighlight
    })
  }
```

**Review by nickoferrall — CHANGES_REQUESTED** [review]

File: packages/client/components/TaskEditor/decorators.ts (hunk 1 above)
Comment (igorlesnenko) [reply packages/client/components/TaskEditor/decorators.ts]: @nickoferrall yeah, this removes thunk, but iteration is still needed. Made it as you suggested.

Comment (nickoferrall) [reply packages/client/components/TaskEditor/decorators.ts]: @igorlesnenko why do we need the while loop as well as `indexOf`? `indexOf` is already iterating through the array?

Comment (nickoferrall) [reply packages/client/components/TaskEditor/decorators.ts]: Ahh OK cool. Thanks, that makes sense. LGTM!

**Comment by mattkrick** [comment]
🤯  wow! great PR @igorlesnenko and great review @nickoferrall
this looks amazing!

**Pull request merged**